### PR TITLE
feat(react-builder): Add a UI builder

### PR DESCRIPTION
**Pull request by levithomason**
# UI Builder Alpha

#### Pull request checklist

- [ ] Addresses an existing issue: Fixes #0000
- [ ] Include a change request file using `$ yarn change`

#### Description of changes

This PR is prototyping a UI builder.  The goal is to provide a tool with which we can validate our components and themes by rapid prototyping interfaces.  A secondary stretch goal is to enable bluelines.

Alpha Features:
- Drag n' Drop components
- Persist UI state to storage
- Shareable links
- Renders content in isolated iframe
- Allows setting props on components

To be done next:
- Save a blueline
- Works for v0 and v7 (hopefully generic)
- Allows theming content
- more stuff

#### Focus areas to test

(optional)

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/OfficeDev/office-ui-fabric-react/pull/12234)

### CODE REVIEW

File: packages/fluentui/react-builder/package.json [hunk 1]
```diff
@@ -0,0 +1,35 @@
+{
+  "name": "@fluentui/react-builder",
+  "description": "An drag and drop UI builder.",
```
Comment (dzearing) [inline packages/fluentui/react-builder/package.json]:
```suggestion
  "description": "A drag and drop UI builder.",
```

File: packages/react-focus/package.json [hunk 2]
```diff
@@ -10,7 +10,7 @@
   ],
   "repository": {
     "type": "git",
-    "url": "https://github.com/microsoft/fluent-ui-react"
+    "url": "https://github.com/OfficeDev/office-ui-fabric-react.git"
```
Comment (miroslavstastny) [inline packages/react-focus/package.json]:
```suggestion
    "url": "https://github.com/microsoft/fluentui.git"
```

File: packages/fluentui/react/package.json [hunk 3]
```diff
@@ -8,6 +8,7 @@
     "@babel/runtime": "^7.7.6",
     "@fluentui/accessibility": "^0.46.0",
     "@fluentui/react-bindings": "^0.46.0",
+    "@fluentui/react-builder": "^0.46.0",
```
Comment (miroslavstastny) [inline packages/fluentui/react/package.json]: Why? No!

File: packages/fluentui/docs/src/index.ejs [hunk 4]
```diff
@@ -24,11 +23,6 @@
       }
     })()
   </script>
-  <style>
-    html {
-      font-size: 14px;
```
Comment (miroslavstastny) [inline packages/fluentui/docs/src/index.ejs]: nit: as Teams theme is 14px-based, having `rem` 14px-based makes debugging easier.

Comment (miroslavstastny) [reply packages/fluentui/docs/src/index.ejs]: reverted this as it causes screener regressions because of rounding

File: packages/fluentui/docs-components/src/CopyToClipboard.tsx [hunk 5]
```diff
@@ -5,13 +5,13 @@ import * as React from 'react';
 export type CopyToClipboardProps = {
   children: (active: boolean, onClick: () => void) => React.ReactElement;
   timeout?: number;
-  value: string;
+  value: string | (() => string);
```
Comment (miroslavstastny) [inline packages/fluentui/docs-components/src/CopyToClipboard.tsx:8]: Added a possibility to only compute the value when needed.

File: packages/fluentui/docs-components/package.json [hunk 6]
```diff
@@ -10,7 +10,9 @@
     "lodash": "^4.17.15",
     "object.values": "^1.1.0",
     "prismjs": "^1.16.0",
-    "prop-types": "^15.7.2"
+    "prop-types": "^15.7.2",
+    "react-ace": "^5.1.2",
+    "react-element-to-jsx-string": "^14.0.2"
```
Comment (layershifter) [inline packages/fluentui/docs-components/package.json:15]: Should we remove these deps from `packages/fluentui/docs/package.json`?

Comment (layershifter) [reply packages/fluentui/docs-components/package.json:15]: Oops, I see, ignore it

File: packages/fluentui/docs/src/index.ejs [hunk 7]
```diff
@@ -5,7 +5,6 @@
   <meta charset="UTF-8">
   <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
   <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.5.0/css/all.css" integrity="sha384-B4dIYHKNBt8Bc12p+WXckhzcICo0wtJAoU8YZTY5qE0Id1GSseTk6S+L3BlXeVIU" crossorigin="anonymous">
-  <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.6.3/css/brands.css" integrity="sha384-1KLgFVb/gHrlDGLFPgMbeedi6tQBLcWvyNUN+YKXbD7ZFbjX6BLpMDf0PJ32XJfX" crossorigin="anonymous">
```
Comment (layershifter) [inline packages/fluentui/docs/src/index.ejs:8]: ❤️

File: packages/fluentui/docs/src/index.ejs [hunk 8]
```diff
@@ -5,7 +5,6 @@
   <meta charset="UTF-8">
   <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
   <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.5.0/css/all.css" integrity="sha384-B4dIYHKNBt8Bc12p+WXckhzcICo0wtJAoU8YZTY5qE0Id1GSseTk6S+L3BlXeVIU" crossorigin="anonymous">
```
Comment (layershifter) [inline packages/fluentui/docs/src/index.ejs:7]:
```suggestion
```

Should we remove the other one, too?

File: packages/fluentui/react-builder/src/components/Anatomy.tsx [hunk 9]
```diff
@@ -0,0 +1,46 @@
+import * as React from 'react';
+import * as FUI from '@fluentui/react-northstar';
+import { ComponentInfo } from '../componentInfo/types';
+
+const ComponentAnatomy = ({ componentInfo }: { componentInfo: ComponentInfo }) => {
```
Comment (layershifter) [inline packages/fluentui/react-builder/src/components/Anatomy.tsx:5]:
```suggestion
const ComponentAnatomy: React.FC<{ componentInfo: ComponentInfo }> = ({ componentInfo }) => {
```

File: packages/fluentui/react-builder/src/components/Anatomy.tsx [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+      <ComponentAnatomy.Slot name="before" />
+      {shorthandProps.map(propDef => (
+        <ComponentAnatomy.Slot key={propDef.name} name={propDef.name} />
+      ))}
+      <ComponentAnatomy.Slot name="after" />
+    </div>
+  );
+};
+
+const ComponentAnatomySlot = ({ name }) => (
```
Comment (layershifter) [inline packages/fluentui/react-builder/src/components/Anatomy.tsx:28]:
```suggestion
const ComponentAnatomySlot: React.FC<{ name: string }> = ({ name }) => (
```

File: packages/fluentui/react-builder/src/components/BrowserWindow.tsx [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+  },
+
+  content: {
+    flex: 1,
+    overflow: 'hidden',
+  },
+};
+
+const BrowserWindow: React.FC<{
+  [key: string]: any;
```
Comment (layershifter) [inline packages/fluentui/react-builder/src/components/BrowserWindow.tsx:92]:
```suggestion
```

File: packages/fluentui/react-builder/src/components/BrowserWindow.tsx [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+  children: React.ReactNode | React.ReactNodeArray;
+  showNavBar: boolean;
+  style?: React.CSSProperties;
+  headerItems?: React.ReactNode | React.ReactNodeArray;
+}> = ({ children, style, showNavBar, headerItems, ...rest }) => (
```
Comment (layershifter) [inline packages/fluentui/react-builder/src/components/BrowserWindow.tsx:97]:
```suggestion
} & React.HTMLAttributes<HTMLDivElement>> = ({ children, style, showNavBar, headerItems, ...rest }) => (
```

if needed

File: packages/fluentui/react-builder/src/components/Canvas.tsx [hunk 13]
```diff
@@ -0,0 +1,311 @@
...
+import Frame, { FrameContextConsumer } from 'react-frame-component';
+
+import { DebugSelector, FiberNavigator, Provider, teamsTheme } from '@fluentui/react-northstar';
+import { JSONTreeElement } from './types';
+import { EventListener } from '@fluentui/react-component-event-listener';
+import { fiberNavFindJSONTreeElement, fiberNavFindOwnerInJSONTree, renderJSONTreeToJSXElement } from '../config';
+import { DebugFrame } from './DebugFrame';
+import { DropSelector } from './DropSelector';
+
+const Canvas = ({
```
Comment (layershifter) [inline packages/fluentui/react-builder/src/components/Canvas.tsx:11]:
```suggestion
const Canvas: React.FC<AllProps> = ({
```

File: packages/fluentui/react-builder/src/components/Canvas.tsx [hunk 14]
```diff
@@ -0,0 +1,311 @@
...
+  onGoToParentComponent,
+  renderJSONTreeElement,
+  style,
+}: {
+  draggingElement: JSONTreeElement;
+  jsonTree: JSONTreeElement;
+  isExpanding?: boolean;
+  isSelecting?: boolean;
+  onDropPositionChange: (dropParent: JSONTreeElement, dropIndex: number) => void;
+  onMouseMove?: ({ clientX, clientY }: { clientX: number; clientY: number }) => void;
```
Comment (layershifter) [inline packages/fluentui/react-builder/src/components/Canvas.tsx:34]:
```suggestion
  onMouseMove?: (e: React.MouseEvent<HTMLDivElement>) => void;
```

Please also other typings and use proper event types

File: packages/fluentui/react-builder/src/components/Canvas.tsx [hunk 15]
```diff
@@ -0,0 +1,311 @@
...
+      `frame-${Math.random()
+        .toString(36)
+        .slice(2)}`,
+    [],
+  );
+
+  const iframeCoordinatesToWindowCoordinates = React.useCallback(
+    (e: MouseEvent) => {
+      const window = (e.target as HTMLElement).ownerDocument.defaultView;
+      const $iframe = window.parent.document.getElementById(iframeId);
```
Comment (layershifter) [inline packages/fluentui/react-builder/src/components/Canvas.tsx:57]: Refs should be more effective

File: packages/fluentui/react-builder/src/components/Canvas.tsx [hunk 16]
```diff
@@ -0,0 +1,311 @@
...
+                .filter(Boolean)
+                .join('\n');
+
+              // console.log(
+              //   element,
+              //   '\nHAS\n',
+              //   { width, height, marginTop, marginRight, marginBottom, marginLeft },
+              //   '\nGETS\n',
+              //   properties,
+              // );
```
Comment (layershifter) [inline packages/fluentui/react-builder/src/components/Canvas.tsx:191]: Should we remove comments in this file?

File: packages/fluentui/react-builder/src/components/Canvas.tsx [hunk 17]
```diff
@@ -0,0 +1,311 @@
...
+    };
+  }, [iframeId, isExpanding, isSelecting]);
+
+  return (
+    <Frame
+      title="Designer Canvas"
+      frameBorder="0"
+      width="100%"
+      height="100%"
+      initialContent='<!DOCTYPE html><html><head><style>html {font-size: 14px;}</style></head><body><div class="frame-root"></div></body></html>'
```
Comment (layershifter) [inline packages/fluentui/react-builder/src/components/Canvas.tsx:236]:
```suggestion
      initialContent='<!DOCTYPE html><html><head><style>html {font-size: VALUE;}</style></head><body><div class="frame-root"></div></body></html>'
```

Should we take `fontSize` value from the main window?

File: packages/fluentui/react-builder/src/components/ComponentTree.tsx [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+      onMoveComponent?.({ clientX: e.clientX, clientY: e.clientY });
+      e.stopPropagation();
+      e.preventDefault();
+    },
+    [onMoveComponent],
+  );
+
+  const handleDelete = React.useCallback(
+    e => {
+      onDeleteComponent?.();
```
Comment (layershifter) [inline packages/fluentui/react-builder/src/components/ComponentTree.tsx:102]: nit, but I definitely _don't like_ `?.()` construction

File: packages/fluentui/react-builder/src/components/DebugButtons.tsx [hunk 19]
```diff
@@ -0,0 +1,125 @@
...
+        cursor: 'pointer',
+      }}
+      {...props}
+    />
+  );
+};
+
+export const TrashDebugButton: React.FunctionComponent<{
+  onClick: React.MouseEventHandler<HTMLButtonElement>;
+}> = props => (
```
Comment (layershifter) [inline packages/fluentui/react-builder/src/components/DebugButtons.tsx:97]:
```suggestion
export const TrashDebugButton: React.FunctionComponent<React.HTMLAttributes<HTMLButtonElement>> = props => (
```

File: packages/fluentui/react-builder/src/components/DebugButtons.tsx [hunk 20]
```diff
@@ -0,0 +1,125 @@
...
+  <DebugButton title="Delete" {...props}>
+    <TrashAltRegularIcon style={{ widht: '1em', height: '1em' }} />
+  </DebugButton>
+);
+
+export const LevelUpDebugButton: React.FunctionComponent<{
+  onClick: React.MouseEventHandler<HTMLButtonElement>;
+}> = props => (
```
Comment (layershifter) [inline packages/fluentui/react-builder/src/components/DebugButtons.tsx:105]:
```suggestion
export const LevelUpDebugButton: React.FunctionComponent<React.HTMLAttributes<HTMLButtonElement>> = props => (
```

File: packages/fluentui/react-builder/src/components/DebugButtons.tsx [hunk 21]
```diff
@@ -0,0 +1,125 @@
...
+  <DebugButton title="Go to parent" {...props}>
+    <LevelUpAltIcon style={{ widht: '1em', height: '1em' }} />
+  </DebugButton>
+);
+
+export const CloneDebugButton: React.FunctionComponent<{
+  onClick: React.MouseEventHandler<HTMLButtonElement>;
+}> = props => (
```
Comment (layershifter) [inline packages/fluentui/react-builder/src/components/DebugButtons.tsx:113]:
```suggestion
export const CloneDebugButton: React.FunctionComponent<React.HTMLAttributes<HTMLButtonElement>> = props => (
```

File: packages/fluentui/react-builder/src/components/DebugButtons.tsx [hunk 22]
```diff
@@ -0,0 +1,125 @@
...
+  <DebugButton title="Clone" {...props}>
+    <CloneIcon style={{ widht: '1em', height: '1em' }} />
+  </DebugButton>
+);
+
+export const MoveDebugButton: React.FunctionComponent<{
+  onClick: React.MouseEventHandler<HTMLButtonElement>;
+}> = props => (
```
Comment (layershifter) [inline packages/fluentui/react-builder/src/components/DebugButtons.tsx:121]:
```suggestion
export const MoveDebugButton: React.FunctionComponent<React.HTMLAttributes<HTMLButtonElement>> = props => (
```

File: packages/fluentui/react-builder/src/components/DebugFrame.tsx [hunk 23]
```diff
@@ -0,0 +1,130 @@
...
+    animationFrameId.current =
+      el.length === 1
+        ? requestAnimationFrame(() => setFramePosition(frameRef.current, el[0]))
+        : requestAnimationFrame(() => hideFrame(frameRef.current));
+
+    return () => cancelAnimationFrame(animationFrameId.current);
+  }, [target, selector, setFramePosition]);
+
+  return (
+    <Ref innerRef={frameRef}>
```
Comment (layershifter) [inline packages/fluentui/react-builder/src/components/DebugFrame.tsx:88]:
```suggestion
```

`frameRef` can be used on `pre` element: `<pre ref={frameRef}>`

File: packages/fluentui/react-builder/src/components/DebugFrame.tsx [hunk 24]
```diff
@@ -0,0 +1,130 @@
...
+    onGoToParent?.();
+  }, [onGoToParent]);
+
+  React.useEffect(() => {
+    if (animationFrameId.current) {
+      cancelAnimationFrame(animationFrameId.current);
+    }
+
+    if (!frameRef.current) {
+      console.log('DebugFrame - no frameRef');
```
Comment (layershifter) [inline packages/fluentui/react-builder/src/components/DebugFrame.tsx:72]: Should we keep logs there

File: packages/fluentui/react-builder/src/components/Description.tsx [hunk 25]
```diff
@@ -0,0 +1,29 @@
+import * as React from 'react';
+import * as FUI from '@fluentui/react-northstar';
+import { ComponentInfo } from '../componentInfo/types';
+import { JSONTreeElement } from './types';
+
+const ComponentDescription = ({
```
Comment (layershifter) [inline packages/fluentui/react-builder/src/components/Description.tsx:6]:
```suggestion
const ComponentDescription: React.FC = ({
```

File: packages/fluentui/react-builder/src/components/Description.tsx [hunk 26]
```diff
@@ -0,0 +1,29 @@
...
+  selectedJSONTreeElement: JSONTreeElement;
+  componentInfo: ComponentInfo;
+}) => {
+  if (!componentInfo) {
+    return null;
+  }
+
+  return (
+    <div>
+      <FUI.Header as="h2" styles={{ display: 'inline-block', margin: 0 }}>
```
Comment (layershifter) [inline packages/fluentui/react-builder/src/components/Description.tsx:19]:
```suggestion
      <Header as="h2" styles={{ display: 'inline-block', margin: 0 }}>
```

File: packages/fluentui/react-builder/src/components/Designer.tsx [hunk 27]
```diff
@@ -0,0 +1,621 @@
...
+  mode: DesignerMode;
+  selectedComponentInfo: ComponentInfo;
+  selectedJSONTreeElement: any;
+  showCode: boolean;
+  code: string | null; // only valid if showCode is set to true
+  codeError: string | null;
+  showJSONTree: boolean;
+};
+
+class Designer extends React.Component<any, DesignerState> {
```
Comment (layershifter) [inline packages/fluentui/react-builder/src/components/Designer.tsx:64]:
```suggestion
class Designer extends React.Component<SomeType, DesignerState> {
```

File: packages/fluentui/react-builder/src/components/Designer.tsx [hunk 28]
```diff
@@ -0,0 +1,621 @@
...
+          width: '100vw',
+          height: '100vh',
+          overflow: 'hidden',
+        }}
+      >
+        {draggingElement && (
+          <>
+            <EventListener type="mousemove" listener={this.handleDrag} target={document} />
+            <EventListener type="mouseup" listener={this.handleDragAbort} target={document} />
+            <Ref innerRef={this.draggingElementRef}>
```
Comment (layershifter) [inline packages/fluentui/react-builder/src/components/Designer.tsx:430]:
```suggestion
```

No need in `Ref` there

File: packages/fluentui/react-builder/src/components/DropSelector.tsx [hunk 29]
```diff
@@ -0,0 +1,189 @@
...
+  filter?;
+  jsonTree: JSONTreeElement;
+  mountDocument?: Document;
+};
+
+type InElementPosition = 'top' | 'right' | 'bottom' | 'left' | 'center';
+
+// left/right wins over top/bottom
+const getInElementPosition: (
+  elementRect: { top: number; left: number; width: number; height: number },
```
Comment (layershifter) [inline packages/fluentui/react-builder/src/components/DropSelector.tsx:20]:
```suggestion
  elementRect: ClientRect,
```
??

File: packages/fluentui/react-builder/src/components/List.tsx [hunk 30]
```diff
@@ -0,0 +1,129 @@
...
+import * as React from 'react';
+import { Box, Menu, Input } from '@fluentui/react-northstar';
+import { SearchIcon } from '@fluentui/react-icons-northstar';
+import { ComponentInfo } from '../componentInfo/types';
+import componentInfoContext from '../componentInfo/componentInfoContext';
+import { EXCLUDED_COMPONENTS } from '../config';
+
+export type ListDisplayModes = 'Display Name' | 'Rendered';
+
+const List = ({
```
Comment (layershifter) [inline packages/fluentui/react-builder/src/components/List.tsx:11]:
```suggestion
const List: React.FC = ({
```

File: packages/fluentui/react-builder/src/components/List.tsx [hunk 31]
```diff
@@ -0,0 +1,129 @@
...
+  onDragStart,
+  style,
+}: {
+  onDragStart?: (componentInfo: ComponentInfo, e: MouseEvent) => any;
```
Comment (layershifter) [inline packages/fluentui/react-builder/src/components/List.tsx:15]:
```suggestion
  onDragStart?: (componentInfo: ComponentInfo, e: React.MouseEvent) => void;
```

File: packages/fluentui/react-builder/src/components/Toolbar.tsx [hunk 32]
```diff
@@ -0,0 +1,115 @@
+import * as React from 'react';
+import { Button, Checkbox, Image, RadioGroup, RadioGroupItemProps } from '@fluentui/react-northstar';
+import { DesignerMode } from './types';
+import { OpenOutsideIcon, TrashCanIcon } from '@fluentui/react-icons-northstar';
+
+const Toolbar = ({
```
Comment (layershifter) [inline packages/fluentui/react-builder/src/components/Toolbar.tsx:6]:
```suggestion
const Toolbar: React.FC = ({
```

File: packages/fluentui/react-builder/src/config.tsx [hunk 33]
```diff
@@ -0,0 +1,721 @@
...
+    } else if (_.isPlainObject(value)) {
+      acc[key] = toJSONTreeElement(value);
+    } else {
+      acc[key] = value;
+    }
+  });
+  return result;
+};
+
+const getUUID = () =>
```
Comment (layershifter) [inline packages/fluentui/react-builder/src/config.tsx:418]: This occurs in multiple files, let's move it to `./utils.ts`

File: packages/fluentui/react-builder/src/config.tsx [hunk 34]
```diff
@@ -0,0 +1,721 @@
...
+    .slice(2);
+
+export const resolveDraggingElement: (displayName: string, draggingElements?) => JSONTreeElement = (
+  displayName,
+  draggingElements = DRAGGING_ELEMENTS,
+) => {
+  const jsonTreeElement = toJSONTreeElement(draggingElements[displayName]);
+  return {
+    uuid: getUUID(),
+    $$typeof: 'Symbol(react.element)',
```
Comment (layershifter) [inline packages/fluentui/react-builder/src/config.tsx:430]:
```suggestion
    $$typeof: Element,
```

`Element` is exported from `react-is`:

image

File: packages/fluentui/react-builder/src/config.tsx [hunk 35]
```diff
@@ -0,0 +1,721 @@
...
+  if (_.isPlainObject(result) && result.$$typeof === 'Symbol(react.element)' && result.uuid) {
+    result.uuid = getUUID();
+  }
+  return result;
+};
+
+/**
+ * Displays a knob with the ability to switch between data `types`.
+ */
+export const MultiTypeKnob: React.FC<{
```
Comment (layershifter) [inline packages/fluentui/react-builder/src/config.tsx:651]: Should be `MultiTypeKnob` in this file?

File: packages/fluentui/react-builder/src/utils/codeToTree.ts [hunk 36]
```diff
@@ -0,0 +1,90 @@
+import { transformSync, types } from '@babel/core';
+import traverse from '@babel/traverse';
+import { parse } from '@babel/parser';
+import reactPreset from '@babel/preset-react';
+import generate from '@babel/generator';
+
+const prefixElementNames = ast => {
```
Comment (layershifter) [inline packages/fluentui/react-builder/src/utils/codeToTree.ts:7]: Let's add types in this file as it's hard to follow there

File: packages/fluentui/react-builder/src/utils/codeToTree.ts [hunk 37]
```diff
@@ -0,0 +1,90 @@
...
+  };
+  traverse(ast, {
+    JSXElement(path) {
+      hackElementName(path.node.openingElement);
+      if (path.node.closingElement) {
+        hackElementName(path.node.closingElement);
+      }
+    },
+  });
+};
```
Comment (layershifter) [inline packages/fluentui/react-builder/src/utils/codeToTree.ts:28]: Actually this looks like a plugin and be done under `transformSync`

File: packages/fluentui/react-builder/src/utils/codeToTree.ts [hunk 38]
```diff
@@ -0,0 +1,90 @@
...
+  //       {
+  //         pragma: 'convert',
+  //         pragmaFrag: '"fragment"',
+  //       },
+  //     ],
+  //   ],
+  // });
+  // console.log('After pluginTransformReactJsx', compiled.code);
+
+  const getUUID = () =>
```
Comment (layershifter) [inline packages/fluentui/react-builder/src/utils/codeToTree.ts:53]: One more `getUUID` 😸

File: packages/fluentui/react-builder/src/utils/codeToTree.ts [hunk 39]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+    const uuid = props?.['data-builder-id'] ?? getUUID();
+    delete props?.['data-builder-id'];
+
+    return {
+      type: name,
+      displayName: name,
+      uuid,
+      ...(name.match(/^[A-Za-z]/) && { $$typeof: 'Symbol(react.element)' }),
```
Comment (layershifter) [inline packages/fluentui/react-builder/src/utils/codeToTree.ts:78]:
```suggestion
      ...(name.match(/^[A-Za-z]/) && { $$typeof: Element }),
```

from `react-is`

File: packages/fluentui/react-northstar/src/components/Slider/Slider.tsx [hunk 40]
```diff
@@ -217,7 +217,8 @@ const Slider: React.FC<WithAsProp<SliderProps>> & FluentComponentStaticProps = p
         ...htmlInputProps,
         as: 'input',
         className: sliderSlotClassNames.input,
-        fluid,
+        // TODO: passing fluid is a bug here, correct? not a valid DOM attr warning in React
+        // fluid,
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Slider/Slider.tsx:221]: Yep, it's a bug, let's remove this a separate PR

File: packages/fluentui/react-builder/src/components/ComponentTree.tsx (hunk 18 above)
Comment (miroslavstastny) [reply packages/fluentui/react-builder/src/components/ComponentTree.tsx:102]: 🤷

File: packages/fluentui/react-builder/src/config.tsx (hunk 35 above)
Comment (miroslavstastny) [reply packages/fluentui/react-builder/src/config.tsx:651]: For now yes, reworked Knobs will be introduced in another PR.

File: packages/fluentui/react-builder/src/components/Canvas.tsx (hunk 14 above)
Comment (miroslavstastny) [reply packages/fluentui/react-builder/src/components/Canvas.tsx:34]: This is not the whole event, just x and y

File: packages/fluentui/react-builder/src/components/Canvas.tsx (hunk 16 above)
Comment (miroslavstastny) [reply packages/fluentui/react-builder/src/components/Canvas.tsx:191]: not yet

File: packages/fluentui/react-builder/src/components/Canvas.tsx (hunk 15 above)
Comment (miroslavstastny) [reply packages/fluentui/react-builder/src/components/Canvas.tsx:57]: Let's handle that in a follow-up PR.

File: packages/fluentui/react-builder/src/config.tsx (hunk 34 above)
Comment (miroslavstastny) [reply packages/fluentui/react-builder/src/config.tsx:430]: I cannot use `Element` as `symbol` - it does not get stringified in `JSON.stringify()` and there are also places where it is not usable (type definition). Let's keep it as is for now and discuss later.

File: packages/fluentui/react-builder/src/components/DropSelector.tsx (hunk 29 above)
Comment (miroslavstastny) [reply packages/fluentui/react-builder/src/components/DropSelector.tsx:20]: `ClientRect` does not have editable `width` and `height` :-(

File: packages/fluentui/react-builder/src/config.tsx [hunk 41]
```diff
@@ -0,0 +1,717 @@
...
+        ) : (
+          types.map(t => (
+            <button key={t} onClick={() => handleChangeType(t)}>
+              {t}
+            </button>
+          ))
+        )}
+      </div>
+      {knob && knob({ options: literalOptions, value, onChange })}
+      {type === 'boolean' && <label> {label}</label>}
```
Comment (jurokapsiar) [inline packages/fluentui/react-builder/src/config.tsx:680]: <label htmlFor={knobId} />

File: packages/fluentui/react-builder/src/config.tsx [hunk 42]
```diff
@@ -0,0 +1,717 @@
...
+    e => setType(e.target.value), // @ts-ignore
+    [],
+  );
+
+  // console.log('MultiTypeKnob', { label, value, type, types });
+
+  return (
+    <div style={{ paddingBottom: '4px', marginBottom: '4px', opacity: knob ? 1 : 0.4 }}>
+      <div>
+        {type !== 'boolean' && <label>{label} </label>}
```
Comment (jurokapsiar) [inline packages/fluentui/react-builder/src/config.tsx:668]: <label htmlFor={knobId} />

File: packages/fluentui/react-builder/src/config.tsx [hunk 43]
```diff
@@ -0,0 +1,717 @@
...
+        )}
+      </div>
+      {knob && knob({ options: literalOptions, value, onChange })}
+      {type === 'boolean' && <label> {label}</label>}
+    </div>
+  );
+};
+
+export const knobs = {
+  boolean: ({ value, onChange }) => (
```
Comment (jurokapsiar) [inline packages/fluentui/react-builder/src/config.tsx:686]: pass a unique id to each knob, use it in `<input id={...} />` so that it can be paired with the label

File: packages/fluentui/react-builder/src/config.tsx [hunk 44]
```diff
@@ -0,0 +1,717 @@
...
+        <option key={opt} value={opt}>
+          {opt}
+        </option>
+      ))}
+    </select>
+  ),
+
+  ReactText: (value, onChange) => knobs.string({ value, onChange }),
+  'React.ElementType': (value, onChange) => knobs.string({ value, onChange }),
+};
```
Comment (jurokapsiar) [inline packages/fluentui/react-builder/src/config.tsx:717]: did you consider handling shorthand as a plain input? It might not work for all cases, but at least it would allow us to add for example `content` to `List.Item`

File: packages/fluentui/react-builder/src/config.tsx (hunk 41 above)
Comment (miroslavstastny) [reply packages/fluentui/react-builder/src/config.tsx:680]: The knobs are temporarily hacked, to be reworked in next iteration.
